grafana/grafana
# Postgres: Switch the datasource plugin from lib/pq to pgx

(note this is the second try to have this merged. first happened in https://github.com/grafana/grafana/pull/81353 but we had to revert it because of a socks-proxy-problem. more about that at the end of the PR)

(fixes https://github.com/grafana/grafana/issues/79671)

for reasons described in the above issue, in the postgres datasource plugin we are switching the connector go module, from `lib/pq` to `pgx`.

NOTE: you can see two snapshot-tests producing changes:
1.  `types_datetime.golden.jsonc` . there is no real change here. if you look closely, you'll see:
    - the json-output below the ascii-table has no changes, so the json-content of the dataframe did not change
    - the change is that the timezone-attribute on the `time.Time` objects changed from `+00` to `UTC`. that's basically the same time zone
2. `timestamp_convert_real.golden.jsonc`. these are real changes. the test deals with cases where we read in a timestamp from the database where in the database it is stored as `REAL`. (32bit-floating-point number).if you compare the values in the `Name: n` column, you can see that the numbers are still very close (exponent matches, and the number also is the same for 6 digits. this problem does not seem avoidable, i think `pgx` uses a different (better) string-representation of the floating-point value, which is the problem. i think this change is acceptable.

what to test:
- all TLS modes, in both file-path and paste-file-content modes
- query editor, does it load all tables, all fields etc?
- do queries work?
- socks proxy

---

# how to test the reason for the "try2":

configuration:
- grafana on local computer
- postgres on local computer (in docker or directly on the compter), available at `localhost:5432`
- we run `microsocks -i 127.0.0.1 -p 5555` , socks proxy on localhost, port 5555

now configure grafana to use the socks proxy. add this to the config:
```yaml
[secure_socks_datasource_proxy]
enabled = true
proxy_address = localhost:5555
allow_insecure = true
````

now run grafana, setup a postgres datasource plugin that connects to the `localhost:5432` postgres through the socks-proxy.
now, watch the output of the `microsocks` app in the terminal:
- when you press [save&test] in grafana:
    - this is what you want to see in the microsocks-terminal: `... connected to localhost:5432`
        - it has to say `localhost:5432`. it must not say `connected to ::1:5432` (or `connected to 127.0.0.1:5432`)

---

NOTE: changes compared to the previous try:
- when the dialer is called, it must receive the unresolved hostname. pgx is resolving the hostname by default, and giving the IP address to the dialer. we do not want it. in other words,  if i say postgres is at `localhost`, i need the `dialer` to be called with  `localhost`, not with `127.0.0.1`. fixed by setting a custom `LookupFn` in `pgx` config.
- `verify-full` and `verify-ca` allows not specifying the root certificate. in that case it fails back to the computer's default trusted authorities (this happens automatically, our code does not have to handle this)
